<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_26) on Mon Mar 19 01:08:36 EDT 2012 -->
<TITLE>
MainCmdline
</TITLE>

<META NAME="date" CONTENT="2012-03-19">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="MainCmdline";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/MainCmdline.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../edu/unc/csbio/RNASim/Config.OrderPreservedProperties.html" title="class in edu.unc.csbio.RNASim"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../edu/unc/csbio/RNASim/PerfectQualityGenerator.html" title="class in edu.unc.csbio.RNASim"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?edu/unc/csbio/RNASim/MainCmdline.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="MainCmdline.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
edu.unc.csbio.RNASim</FONT>
<BR>
Class MainCmdline</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../resources/inherit.gif" ALT="extended by "><B>edu.unc.csbio.RNASim.MainCmdline</B>
</PRE>
<HR>
<DL>
<DT><PRE>public class <B>MainCmdline</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
The class <code>MainCmdline</code> is the main class
<P>

<P>
<DL>
<DT><B>Version:</B></DT>
  <DD>0.1</DD>
<DT><B>Author:</B></DT>
  <DD>Shunping Huang <sphuang@cs.unc.edu>, Jack Wang <zhew@live.unc.edu></DD>
</DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;java.util.HashMap&lt;java.lang.String,java.lang.Double&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#abundancePool">abundancePool</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;a Hashmap that maps a string to an abundance value</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#annotLimit">annotLimit</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;max number of lines allowed for annotation file</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#chrom">chrom</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this will be the prefix of the all read ID</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#chromMatching">chromMatching</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;specify the matching of chromosome string, Exact or Fuzzy.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;java.util.TreeSet&lt;java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#chromPool">chromPool</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;a TreeSet that stores the chromosome names</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#coverageFactor">coverageFactor</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;coverage factor = total read length / total sequence size</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#distMean">distMean</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;distribution mean value, will be used when calculating abundance.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#expressionPercentage">expressionPercentage</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;percentage of gene that will be expressed</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#fragMaxLength">fragMaxLength</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;maximum sequence length</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#fragMinLength">fragMinLength</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;minimum sequence length</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;java.util.TreeMap&lt;java.lang.String,<A HREF="../../../../edu/unc/csbio/data/Gene.html" title="class in edu.unc.csbio.data">Gene</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#genePool">genePool</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;a TreeMap that maps a string to a gene</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;<A HREF="../../../../edu/unc/csbio/data/Gene.html" title="class in edu.unc.csbio.data">Gene</A>[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#genes">genes</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;an array that stores the genes</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#IDPREFIX">IDPREFIX</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this will be the prefix of the all read ID</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#inputBufferSize">inputBufferSize</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;input buffer size</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;<A HREF="../../../../edu/unc/csbio/util/FqWriter.html" title="class in edu.unc.csbio.util">FqWriter</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#out1">out1</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;the file location of fq writer file (for first pair end read)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;<A HREF="../../../../edu/unc/csbio/util/FqWriter.html" title="class in edu.unc.csbio.util">FqWriter</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#out2">out2</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;the file location of fq writer file (for second pair end read)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#outputBufferSize">outputBufferSize</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;output buffer size</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#processBufferSize">processBufferSize</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;buffer size for accumulating gene range</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;<A HREF="../../../../edu/unc/csbio/RNASim/QualityGenerator.html" title="class in edu.unc.csbio.RNASim">QualityGenerator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#qg">qg</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;a qualityGenerator</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#readLength">readLength</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;read length</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#startTime">startTime</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;time when program start running</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;java.util.HashMap&lt;java.lang.String,java.util.ArrayList&lt;<A HREF="../../../../edu/unc/csbio/data/Feature.html" title="class in edu.unc.csbio.data">Feature</A>&gt;&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#svPool">svPool</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;an array that stores the structure variants</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#tDistDOF">tDistDOF</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;the value used to calculate distMean.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#totalGenes">totalGenes</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;total number of the genes</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#totalGenesByChrom">totalGenesByChrom</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;total number of genes in a chromosome</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#totalReads">totalReads</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;total number of the reads</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#totalReadsByChrom">totalReadsByChrom</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;total number of reads in a chromosome</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#totalTranscripts">totalTranscripts</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;total number of transcripts</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#totalTranscriptsByChrom">totalTranscriptsByChrom</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;total number of transcripts in a chromosome</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) static&nbsp;java.util.TreeMap&lt;java.lang.String,<A HREF="../../../../edu/unc/csbio/data/Transcript.html" title="class in edu.unc.csbio.data">Transcript</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#transcriptPool">transcriptPool</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;a TreeMap that maps a string to a transcript</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#MainCmdline()">MainCmdline</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#abort(java.lang.Exception)">abort</A></B>(java.lang.Exception&nbsp;e)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;the method to abort the program</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#BuildAbundance(java.lang.String)">BuildAbundance</A></B>(java.lang.String&nbsp;abundanceFn)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this method creates abundance file with default settings</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#GenerateSequence()">GenerateSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this method generate read sequence from given <code>Config</code> file</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#GenerateSequenceForGenes(java.lang.String)">GenerateSequenceForGenes</A></B>(java.lang.String&nbsp;chrom)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given a chromosome, this method call
 <code>public static void GenerateSequenceForSingleGene(Gene g,
                        SequenceBuffer buffer) throws IOException</code> on all genes of that
 chromosome, and prints out total genes, reads and transcripts associated
 with chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#GenerateSequenceForSingleGene(edu.unc.csbio.data.Gene, edu.unc.csbio.data.SequenceBuffer)">GenerateSequenceForSingleGene</A></B>(<A HREF="../../../../edu/unc/csbio/data/Gene.html" title="class in edu.unc.csbio.data">Gene</A>&nbsp;g,
                              <A HREF="../../../../edu/unc/csbio/data/SequenceBuffer.html" title="class in edu.unc.csbio.data">SequenceBuffer</A>&nbsp;buffer)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;given a gene, this method generates read and appends read information to
 output fq file</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#main(java.lang.String[])">main</A></B>(java.lang.String[]&nbsp;args)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this is the driver for the whole program</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#printRuntime()">printRuntime</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;the method to print out the run time of the program</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#ReadAnnotation()">ReadAnnotation</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this method reads annotation file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#ReadConfig(java.lang.String[])">ReadConfig</A></B>(java.lang.String[]&nbsp;args)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this method read configuration file</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#ReadStructureVariant()">ReadStructureVariant</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this method read structure variants.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#ReadTranscriptAbundance()">ReadTranscriptAbundance</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this method read the transcript abundance from
 <code>Abundance_File</code> in <code>Config</code> file</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../edu/unc/csbio/RNASim/MainCmdline.html#shuffle(java.lang.Object[])">shuffle</A></B>(java.lang.Object[]&nbsp;array)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;given an array, this method sequentially changes a position with another
 randomly selected position, for n times, where n is the size of array.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="IDPREFIX"><!-- --></A><H3>
IDPREFIX</H3>
<PRE>
static java.lang.String <B>IDPREFIX</B></PRE>
<DL>
<DD>this will be the prefix of the all read ID
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="chrom"><!-- --></A><H3>
chrom</H3>
<PRE>
static java.lang.String <B>chrom</B></PRE>
<DL>
<DD>this will be the prefix of the all read ID
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="chromMatching"><!-- --></A><H3>
chromMatching</H3>
<PRE>
static java.lang.String <B>chromMatching</B></PRE>
<DL>
<DD>specify the matching of chromosome string, Exact or Fuzzy.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="annotLimit"><!-- --></A><H3>
annotLimit</H3>
<PRE>
static int <B>annotLimit</B></PRE>
<DL>
<DD>max number of lines allowed for annotation file
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="inputBufferSize"><!-- --></A><H3>
inputBufferSize</H3>
<PRE>
static int <B>inputBufferSize</B></PRE>
<DL>
<DD>input buffer size
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="outputBufferSize"><!-- --></A><H3>
outputBufferSize</H3>
<PRE>
static int <B>outputBufferSize</B></PRE>
<DL>
<DD>output buffer size
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="processBufferSize"><!-- --></A><H3>
processBufferSize</H3>
<PRE>
static int <B>processBufferSize</B></PRE>
<DL>
<DD>buffer size for accumulating gene range
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="readLength"><!-- --></A><H3>
readLength</H3>
<PRE>
static int <B>readLength</B></PRE>
<DL>
<DD>read length
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="fragMinLength"><!-- --></A><H3>
fragMinLength</H3>
<PRE>
static int <B>fragMinLength</B></PRE>
<DL>
<DD>minimum sequence length
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="fragMaxLength"><!-- --></A><H3>
fragMaxLength</H3>
<PRE>
static int <B>fragMaxLength</B></PRE>
<DL>
<DD>maximum sequence length
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="coverageFactor"><!-- --></A><H3>
coverageFactor</H3>
<PRE>
static double <B>coverageFactor</B></PRE>
<DL>
<DD>coverage factor = total read length / total sequence size
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="expressionPercentage"><!-- --></A><H3>
expressionPercentage</H3>
<PRE>
static double <B>expressionPercentage</B></PRE>
<DL>
<DD>percentage of gene that will be expressed
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="distMean"><!-- --></A><H3>
distMean</H3>
<PRE>
static double <B>distMean</B></PRE>
<DL>
<DD>distribution mean value, will be used when calculating abundance.
 abundance is related to distMean and probability associated with
 T-distribution.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="tDistDOF"><!-- --></A><H3>
tDistDOF</H3>
<PRE>
static double <B>tDistDOF</B></PRE>
<DL>
<DD>the value used to calculate distMean. distMean = the value of
 TDistribution at tDistDOF
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="startTime"><!-- --></A><H3>
startTime</H3>
<PRE>
static long <B>startTime</B></PRE>
<DL>
<DD>time when program start running
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="transcriptPool"><!-- --></A><H3>
transcriptPool</H3>
<PRE>
static java.util.TreeMap&lt;java.lang.String,<A HREF="../../../../edu/unc/csbio/data/Transcript.html" title="class in edu.unc.csbio.data">Transcript</A>&gt; <B>transcriptPool</B></PRE>
<DL>
<DD>a TreeMap that maps a string to a transcript
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="genePool"><!-- --></A><H3>
genePool</H3>
<PRE>
static java.util.TreeMap&lt;java.lang.String,<A HREF="../../../../edu/unc/csbio/data/Gene.html" title="class in edu.unc.csbio.data">Gene</A>&gt; <B>genePool</B></PRE>
<DL>
<DD>a TreeMap that maps a string to a gene
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="abundancePool"><!-- --></A><H3>
abundancePool</H3>
<PRE>
static java.util.HashMap&lt;java.lang.String,java.lang.Double&gt; <B>abundancePool</B></PRE>
<DL>
<DD>a Hashmap that maps a string to an abundance value
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="chromPool"><!-- --></A><H3>
chromPool</H3>
<PRE>
static java.util.TreeSet&lt;java.lang.String&gt; <B>chromPool</B></PRE>
<DL>
<DD>a TreeSet that stores the chromosome names
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="svPool"><!-- --></A><H3>
svPool</H3>
<PRE>
static java.util.HashMap&lt;java.lang.String,java.util.ArrayList&lt;<A HREF="../../../../edu/unc/csbio/data/Feature.html" title="class in edu.unc.csbio.data">Feature</A>&gt;&gt; <B>svPool</B></PRE>
<DL>
<DD>an array that stores the structure variants
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="genes"><!-- --></A><H3>
genes</H3>
<PRE>
static <A HREF="../../../../edu/unc/csbio/data/Gene.html" title="class in edu.unc.csbio.data">Gene</A>[] <B>genes</B></PRE>
<DL>
<DD>an array that stores the genes
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="out1"><!-- --></A><H3>
out1</H3>
<PRE>
static <A HREF="../../../../edu/unc/csbio/util/FqWriter.html" title="class in edu.unc.csbio.util">FqWriter</A> <B>out1</B></PRE>
<DL>
<DD>the file location of fq writer file (for first pair end read)
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="out2"><!-- --></A><H3>
out2</H3>
<PRE>
static <A HREF="../../../../edu/unc/csbio/util/FqWriter.html" title="class in edu.unc.csbio.util">FqWriter</A> <B>out2</B></PRE>
<DL>
<DD>the file location of fq writer file (for second pair end read)
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="qg"><!-- --></A><H3>
qg</H3>
<PRE>
static <A HREF="../../../../edu/unc/csbio/RNASim/QualityGenerator.html" title="class in edu.unc.csbio.RNASim">QualityGenerator</A> <B>qg</B></PRE>
<DL>
<DD>a qualityGenerator
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="totalReads"><!-- --></A><H3>
totalReads</H3>
<PRE>
static long <B>totalReads</B></PRE>
<DL>
<DD>total number of the reads
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="totalGenes"><!-- --></A><H3>
totalGenes</H3>
<PRE>
static long <B>totalGenes</B></PRE>
<DL>
<DD>total number of the genes
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="totalTranscripts"><!-- --></A><H3>
totalTranscripts</H3>
<PRE>
static long <B>totalTranscripts</B></PRE>
<DL>
<DD>total number of transcripts
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="totalReadsByChrom"><!-- --></A><H3>
totalReadsByChrom</H3>
<PRE>
static long <B>totalReadsByChrom</B></PRE>
<DL>
<DD>total number of reads in a chromosome
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="totalTranscriptsByChrom"><!-- --></A><H3>
totalTranscriptsByChrom</H3>
<PRE>
static long <B>totalTranscriptsByChrom</B></PRE>
<DL>
<DD>total number of transcripts in a chromosome
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="totalGenesByChrom"><!-- --></A><H3>
totalGenesByChrom</H3>
<PRE>
static long <B>totalGenesByChrom</B></PRE>
<DL>
<DD>total number of genes in a chromosome
<P>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="MainCmdline()"><!-- --></A><H3>
MainCmdline</H3>
<PRE>
public <B>MainCmdline</B>()</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="printRuntime()"><!-- --></A><H3>
printRuntime</H3>
<PRE>
public static void <B>printRuntime</B>()</PRE>
<DL>
<DD>the method to print out the run time of the program
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="abort(java.lang.Exception)"><!-- --></A><H3>
abort</H3>
<PRE>
public static void <B>abort</B>(java.lang.Exception&nbsp;e)</PRE>
<DL>
<DD>the method to abort the program
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="shuffle(java.lang.Object[])"><!-- --></A><H3>
shuffle</H3>
<PRE>
public static void <B>shuffle</B>(java.lang.Object[]&nbsp;array)</PRE>
<DL>
<DD>given an array, this method sequentially changes a position with another
 randomly selected position, for n times, where n is the size of array.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="ReadConfig(java.lang.String[])"><!-- --></A><H3>
ReadConfig</H3>
<PRE>
public static void <B>ReadConfig</B>(java.lang.String[]&nbsp;args)</PRE>
<DL>
<DD>this method read configuration file
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>args</CODE> - the argument array from command line</DL>
</DD>
</DL>
<HR>

<A NAME="ReadAnnotation()"><!-- --></A><H3>
ReadAnnotation</H3>
<PRE>
public static void <B>ReadAnnotation</B>()</PRE>
<DL>
<DD>this method reads annotation file.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="BuildAbundance(java.lang.String)"><!-- --></A><H3>
BuildAbundance</H3>
<PRE>
public static void <B>BuildAbundance</B>(java.lang.String&nbsp;abundanceFn)</PRE>
<DL>
<DD>this method creates abundance file with default settings
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>abundanceFn</CODE> - file location of abundance file to be created</DL>
</DD>
</DL>
<HR>

<A NAME="ReadTranscriptAbundance()"><!-- --></A><H3>
ReadTranscriptAbundance</H3>
<PRE>
public static void <B>ReadTranscriptAbundance</B>()</PRE>
<DL>
<DD>this method read the transcript abundance from
 <code>Abundance_File</code> in <code>Config</code> file
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="ReadStructureVariant()"><!-- --></A><H3>
ReadStructureVariant</H3>
<PRE>
public static void <B>ReadStructureVariant</B>()</PRE>
<DL>
<DD>this method read structure variants.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="GenerateSequenceForSingleGene(edu.unc.csbio.data.Gene, edu.unc.csbio.data.SequenceBuffer)"><!-- --></A><H3>
GenerateSequenceForSingleGene</H3>
<PRE>
public static void <B>GenerateSequenceForSingleGene</B>(<A HREF="../../../../edu/unc/csbio/data/Gene.html" title="class in edu.unc.csbio.data">Gene</A>&nbsp;g,
                                                 <A HREF="../../../../edu/unc/csbio/data/SequenceBuffer.html" title="class in edu.unc.csbio.data">SequenceBuffer</A>&nbsp;buffer)
                                          throws java.io.IOException</PRE>
<DL>
<DD>given a gene, this method generates read and appends read information to
 output fq file
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>g</CODE> - A <code>Gene</code> object<DD><CODE>buffer</CODE> - A <code>SequenceBuffer</code> object
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if an input or output exception occurred</DL>
</DD>
</DL>
<HR>

<A NAME="GenerateSequenceForGenes(java.lang.String)"><!-- --></A><H3>
GenerateSequenceForGenes</H3>
<PRE>
public static void <B>GenerateSequenceForGenes</B>(java.lang.String&nbsp;chrom)
                                     throws java.io.IOException</PRE>
<DL>
<DD>Given a chromosome, this method call
 <code>public static void GenerateSequenceForSingleGene(Gene g,
                        SequenceBuffer buffer) throws IOException</code> on all genes of that
 chromosome, and prints out total genes, reads and transcripts associated
 with chromosome.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>chrom</CODE> - A chromosome
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if an input or output exception occurred</DL>
</DD>
</DL>
<HR>

<A NAME="GenerateSequence()"><!-- --></A><H3>
GenerateSequence</H3>
<PRE>
public static void <B>GenerateSequence</B>()</PRE>
<DL>
<DD>this method generate read sequence from given <code>Config</code> file
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="main(java.lang.String[])"><!-- --></A><H3>
main</H3>
<PRE>
public static void <B>main</B>(java.lang.String[]&nbsp;args)</PRE>
<DL>
<DD>this is the driver for the whole program
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>args</CODE> - program arguments</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/MainCmdline.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../edu/unc/csbio/RNASim/Config.OrderPreservedProperties.html" title="class in edu.unc.csbio.RNASim"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../edu/unc/csbio/RNASim/PerfectQualityGenerator.html" title="class in edu.unc.csbio.RNASim"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?edu/unc/csbio/RNASim/MainCmdline.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="MainCmdline.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
